Nina-W10 implement ioctl and fix timeout.

iabdalkader opened the pull request:
* Add function to check socket state (readable/accept ready)
* Implement ioctl (tested select.select)
* Fix timeout handling to match modusocket (0==non-block, -1=blocking, else timeout in ms).
* Fix setsockopt null nic_type reference.
* Add makefile and common socket options.
* Close accepted sockets when `multi_net` tests are done.
* Fix `multi_net/udp_data.py` bind on non-reusable socket.
* Create new sockets in blocking mode.

iabdalkader committed:
drivers/ninaw10/nina_wifi_drv.c: Add function to check socket state.

iabdalkader committed:
extmod/network_ninaw10.c: Implement ioctl.

* Note AFAIK there's no function to query if sock is writable.

iabdalkader committed:
ninaw10: Fix timeout handling.

* Change timeout meaning to match modusocket.

iabdalkader commented:
@robert-hh Can you please run your tests again with this PR and let me know. I made some fixes to timeout and implemented ioctl for select, it should be working now.

dpgeorge commented:
You can try running the network multitests against 2x boards with WiFi (eg Nano Connect and PYBD).  Like this:
```
$ cd tests
$ ./run-multitests.py -i pyb:a0 -i pyb:a1 multi_net/*.py
```

robert-hh commented:
@iabdalkader A simple test of select worked.
@dpgeorge multitest.py fails all tests. Error messages complain about rp2 not supporting SOL_SOCKET or ussl.

dpgeorge commented:
`tests/multi_net/udp_data.py` is pretty simple and doesn't need `SOL_SOCKET`, so I'm surprised that fails.

`tests/multi_net/tcp_data.py` should be made to work, it would need at least `s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)` supported.  If this test doesn't work then there's no way `asyncio` will work with sockets.

iabdalkader commented:
> `tests/multi_net/udp_data.py` is pretty simple and doesn't need `SOL_SOCKET`, so I'm surprised that fails.

It fails because of a timeout, same with tcp_data a timeout, if I set the socket blocking they both pass. The issue is, the driver used `0` for blocking so by default the sockets were blocking. In this PR I changed this behavior to align with `modusocket` timeout behavior, so now by default sockets are non-blocking.

iabdalkader committed:
extmod/modusocket.c: Fix setsockopt null nic_type reference.

iabdalkader commented:
Is there a way to know which device is running which test ? It gets very confusing when resetting devices. Also I found out that I can just run the test against micropython itself from unix port. Side note on my machine unix port fails to build because of something about sha1 headers, if I switch it to mbdtls it builds. This could be an issue with my local repo.

I found a bug in modusocket, `socket_setsockopt()` is called on unbound sockets (NULL nic_type) most unit-tests do this:
```Python
s = socket.socket()
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)  # <- nic is NULL
```
In `socket_setsockopt()` there's No check for `self->nic_type == NULL`. I pushed another commit to fix it, without raising an error to avoid breaking the unit-tests. Let me know if I should raise an error and fix the unit-tests. Also in `tcp_data.py` the accepted socket `s2` is not closed.

iabdalkader committed:
extmod/modusocket.c: Add makefile and common socket options.

dpgeorge commented:
> Note by default sockets are created in blocking mode in Python [see docs](https://docs.python.org/3/library/socket.html#notes-on-socket-timeouts) but `modusocket` creates non-blocking sockets by default.

`modusocket` must create non-blocking sockets by default.  It must conform to CPython (and other MicroPython ports) for that behaviour.

> Is there a way to know which device is running which test ?

For the multitests?  It prints the device (or executable if running on the PC) that an instance runs on.  If you're hard-resetting devices then you need to keep track of which comes up at which port.  Note you can pass the `-t` option to `run-multitests.py` to trace the output.  It would be possible to augment that tracing output by additional info about which instance is running on what device.

> on my machine unix port fails to build because of something about sha1 headers

It builds OK for me, and builds under CI.  Make sure you sync and update the axtls submodule (the repo changed recently).

> Also in `tcp_data.py` the accepted socket `s2` is not closed.

Devices should be tolerant to sockets not being closed, eg `gc_sweep_all()` should clean up all open sockets via the finaliser.  That said, tests should cleanly close/free resources they use, so that should be fixed in `tcp_data.py`.

iabdalkader committed:
tests/multi_net: Close accepted sockets when tests are done.

* gc_sweep_all() cleans up sockets via the finaliser,
    but tests should cleanly free resources they use.

iabdalkader commented:
> It must conform to CPython (and other MicroPython ports) for that behaviour.

But CPython creates **blocking** sockets by default, from the [docs](https://docs.python.org/3/library/socket.html#notes-on-socket-timeouts):

>  Sockets are by default always created in blocking mode, but this can be changed by calling setdefaulttimeout().

And :
```Python
Python 3.10.1 (main, Dec 18 2021, 23:53:45) [GCC 11.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket, platform
>>> platform.python_implementation()
'CPython'
>>> s = socket.socket()
>>> s.getblocking()
True
```

That said `network_nina.c` just follows MicroPython's modusocket behavior by leaving sockets timeout unchanged when bound. However, this means it can't pass most of the tests in multi_net because the sockets are non-blocking and they should timeout.

>>  That said, tests should cleanly close/free resources they use, so that should be fixed in tcp_data.py.

Fixed, found 2 more files doing the same.

dpgeorge commented:
> That said `network_nina.c` just follows MicroPython's modusocket behavior by leaving sockets timeout unchanged when bound. However, this means it can't pass most of the tests in multi_net because the sockets are non-blocking and they should timeout

The tests are a good "source of truth" and need to be trusted.  I need to look into it more but it sounds like the `modusocket` behaviour needs to be changed.  `extmod/modlwip.c` should be doing things correctly, it can pass the tests.

iabdalkader commented:
> `extmod/modlwip.c` should be doing things correctly, it can pass the tests.

It does Not use modusocket, must be creating blocking sockets by default.

iabdalkader commented:
> . I need to look into it more but it sounds like the `modusocket` behaviour needs to be changed

Can you please take a look at `socket_read()` too ? Should this function always succeed and read the requested bytes when called ? When I tested `webrepl` I had to make this function blocking, otherwise weprepl fails if it times out. I expected the upper layers to poll for readability first before calling `socket_read()`, but that doesn't seem to be the case.

dpgeorge commented:
> Can you please take a look at `socket_read()` too ? Should this function always succeed and read the requested bytes when called ?

It should respect the configured timeout, so by default is blocking and wait forever for at least one byte (or disconnection).  It doesn't need to read all the requested bytes, just at least one byte and at most the requested number of bytes.

iabdalkader commented:
> It should respect the configured timeout, so by default is blocking

Okay I'm going to switch `modusocket` to create blocking `timeout=-1` sockets by default. Also another issue with the tests, rp2 does Not use internal error numbers.

iabdalkader committed:
tests/multi_net/udp_data.py: Allow reusing port before bind.

iabdalkader committed:
extmod/modusocket.c: Create new sockets in blocking mode.

* To conform with CPython and other MicroPython ports.

dpgeorge commented:
The `ssl_data` test should pass since 2c9dc5742abf1f3c787f166d2cbffb4b556ff77b and 5e506567a390db1b1d2df39ae192f39ed9f617c3

dpgeorge commented:
@iabdalkader are you happy with this PR as-is?  Would be good to get this set of changes in now, and then work on further improvements later, to make more of the network tests succeed on rp2.

> What's required to support `asyncio` ?

I'll have a look after this is merged.

dpgeorge commented on extmod/modnetwork.h:
```diff
@@ -38,6 +38,20 @@
+#define MOD_NETWORK_SO_RCVBUF       (0x1002)
+#define MOD_NETWORK_SO_SNDTIMEO     (0x1005)
+#define MOD_NETWORK_SO_RCVTIMEO     (0x1006)
+#define MOD_NETWORK_SO_NO_CHECK     (0x100a)
```
I'm not sure all these constants are needed (yet).  `modlwip.c` only provides `SOL_SOCKET` and `SO_REUSEADDR` and those seem to be enough so far.

iabdalkader committed:
ninaw10: Return standard error numbers.

iabdalkader commented:
> The ssl_data test should pass since 2c9dc57 and 5e50656

Note I rebased on master, and tested with `PYBD` again (along with changes in this PR) 2 tests still fail:

```bash
./run-multitests.py -i pyb:a0 multi_net/*
....
16 tests performed
14 tests passed
2 tests failed: multi_net/tcp_client_rst.py multi_net/uasyncio_tcp_client_rst.py
```

> I'm not sure all these constants are needed (yet). modlwip.c only provides SOL_SOCKET and SO_REUSEADDR and those seem to be enough so far.

I removed some of the rarely used options. But FYI I requested some fixes to the Nina firmware and features, hopefully soon there will be functions to set socket options, set non-blocking flag, set socket timeouts (on socket level) and return `errno`.

> are you happy with this PR as-is? Would be good to get this set of changes in now, and then work on further improvements later, to make more of the network tests succeed on rp2.

Yes, I pushed another commit to use MP error numbers. Note right now only the `async_io` tests fails, in addition to `tcp_client_rst.py` and `uasyncio_tcp_client_rst.py` (that fail on PYBD too) and `udp_data.py` (when running as instance0).

dpgeorge commented:
Ok, great.  The `tcp_client_rst` tests are a bit brutal and lwip currently doesn't handle TCP RST like unix does.  They are ok to fail at this stage.

dpgeorge commented:
I'm now noticing that the Nina driver does not retain connection on soft-reset.  So in order to run the network tests I made a boot.py which connects to the local WLAN on boot (which is slow).

Other ports like stm32, esp32, esp8266 all retain WLAN connection on soft reset.  I think it'd make it much more user friendly if this were the case for Nina.  Since the Nina co-processor (an esp32) is not reset on soft-reset it should be relatively easy to do this.

iabdalkader commented:
> I'm now noticing that the Nina driver does not retain connection on soft-reset

It does retain the connection, but I think it should check and short-cut the connection if it's connected already.

```Python
=== wlan = network.WLAN(network.STA_IF)
=== wlan.active(True)
=== wlan.connect(SSID, KEY)
>>> wlan.ifconfig()
('192.168.1.103', '255.255.255.0', '192.168.1.1', '192.168.1.1')
>>> 
MPY: soft reboot
MicroPython v1.18-19-ga7b6fd45e on 2022-01-20; Arduino Nano RP2040 Connect with RP2040
Type "help()" for more information.
>>> import network
>>> wlan = network.WLAN(network.STA_IF)
>>> wlan.ifconfig()
('192.168.1.103', '255.255.255.0', '192.168.1.1', '192.168.1.1')
```

iabdalkader commented:
Should I return if it's already connected ?

dpgeorge commented:
> Should I return if it's already connected ?

Yes.

dpgeorge commented:
> > Should I return if it's already connected ?
> 
> Yes.

Actually, no, I think it should stay like it is and reattempt the connect.

iabdalkader commented:
> Actually, no, I think it should stay like it is and reattempt the connect.

Okay, no problem. But I realized that `active()` re-initializes the module every time, that can be fixed.

dpgeorge commented:
I realised that the NIC is not registered if the WLAN instance is not created, so now my `boot.py` contains `network.WLAN()` which is enough to get it working (once it's already connected to the AP).

But if I run a test back-to-back then it fails (this is the failure after it already succeeding once before):
```
$ ./run-multitests.py -i pyb:a0 -t multi_net/tcp_data.py
multi_net/tcp_data.py on ttyACM0|micropython: 
TRACE ttyACM0|micropython:
   254 i0 : SET IP = '192.168.20.38'
   258 i0 : 
   258 i0 : Traceback (most recent call last):
  File "<stdin>", line 76, in <module>
  File "<stdin>", line 12, in instance0
OSError: -1
```

iabdalkader commented:
> But if I run a test back-to-back then it fails

I see an issue to, reverted my last commit.. Could it be the pin state ? some pins control BLE/WiFi mode.

dpgeorge commented:
> reverted my last commit..

I tested without your most recent commit (to not activate if already active) and the issue is there.

iabdalkader commented:
> > reverted my last commit..
> 
> I tested without your most recent commit (to not activate if already active) and the issue is there.

Oh I see the issue, it fails to bind the socket to the same addr because currently there's no way to set `SO_REUSEADDR` flag. If the module is reset, it shouldn't fail. So `boot.py` must have a `wlan.active(True)`. I can run `tcp_data.py` multiple times in a row with no issue.

dpgeorge commented:
> Oh I see the issue, it fails to bind the socket to the same addr because currently there's no way to set `SO_REUSEADDR` flag

Is that because the Nina FW doesn't support it yet?

iabdalkader commented:
> Is that because the Nina FW doesn't support it yet?

Yes, there's no command to call `setsockopt` and it's missing a few other things. I put a request for fixes & features. Once it's implemented I will revisit the driver and tests.

dpgeorge commented:
Ok.

I suggest we keep that most recent commit, to not reactivate if it's already active.  So this PR looks good to go.

iabdalkader committed:
extmod/network_ninaw10.c: Retain module state on soft-reboot.

iabdalkader commented:
Done. However, note if pin state/function change on soft-reboot it may be an issue.

dpgeorge commented:
Pin IO levels don't change on soft-reset.  But their IRQ handlers are all deinit'd.

iabdalkader commented:
> Pin IO levels don't change on soft-reset. But their IRQ handlers are all deinit'd.

It doesn't use pin IRQs so I guess this is good, but also note bind() on same addr will fail if the module is not reinitialized on soft-reboot, which may be confusing to users. This will be fixed eventually but right now I don't recommend merging the last commit :sweat_smile:

dpgeorge commented:
> note bind() on same addr will fail if the module is not reinitialized on soft-reboot, which may be confusing to users. This will be fixed eventually but right now I don't recommend merging the last commit sweat_smile

Ah, I understand.  Ok I will drop that last commit.

dpgeorge commented on drivers/ninaw10/nina_wifi_drv.c:
```diff
@@ -752,23 +753,37 @@ int nina_socket_listen(int fd, uint32_t backlog) {
-
-        if (timeout && (mp_hal_ticks_ms() - start) >= timeout) {
-            return NINA_ERROR_TIMEOUT;
+        if (timeout == 0 || (timeout > 0 && (mp_hal_ticks_ms() - start) >= timeout)) {
```
It looks like this returns -MP_ETIMEDOUT even if there is data available: if the above call to `nina_socket_avail` returns non-zero in `sock` (which should be called `avail`) then that's not checked until the next for loop iteration, and if `timeout==0` that will never occur.

dpgeorge replied:
And below in `NINA_CMD_SOCKET_REMOTE_ADDR`, is `sock` supposed to be passed as the first argument?  Below it seems to be used as the new fd...

iabdalkader replied:
Fixed.

iabdalkader replied:
> nina_socket_avail returns non-zero in sock (which should be called avail)

What should be called `avail` ?

> And below in NINA_CMD_SOCKET_REMOTE_ADDR, is sock supposed to be passed as the first argument? Below it seems to be used as the new fd...

Yes, it returns the remote IP/PORT, and it is the accepted socket.

dpgeorge replied:
But what about `sock` below, it's now the number of bytes available on the socket, when it should be the new file descriptor of the newly accepted socket.

iabdalkader replied:
`nina_socket_avail()` or this command `NINA_CMD_SOCKET_AVAIL`, returns the number of bytes available or a newly accepted socket if the parent socket is in listen state.

iabdalkader replied:
Then, `sock` is used in `NINA_CMD_SOCKET_REMOTE_ADDR` to return the accepted socket's IP and port.

iabdalkader replied:
It's confusing, but it's the way the protocol is designed, it could've been much better for sure. It took me a while to figure out some of the vague commands inner workings by looking at the firmware source.

dpgeorge replied:
ok!

dpgeorge commented on extmod/network_ninaw10.c:
```diff
@@ -536,17 +536,31 @@ STATIC int network_ninaw10_socket_setsockopt(mod_network_socket_obj_t *socket, m
+    if (request == MP_STREAM_POLL) {
+        if (arg & MP_STREAM_POLL_RD) {
+            uint16_t avail = 0;
+            if (nina_socket_avail(socket->fileno, socket->type, &avail) != 0) {
```
I'm not sure `socket->type` has the correct value to pass to `nina_socket_avail`, it needs to convert, eg, `MOD_NETWORK_SOCK_STREAM` to `NINA_SOCKET_TYPE_TCP`.

iabdalkader replied:
Good find! It doesn't cause an issue but I fixed it anyway.

dpgeorge commented:
Merged in 981664fd079a379dfd9749e9068781590813b401 through 1aac151d6894cf892c0ff7bb38e5589580d6c080

iabdalkader commented:
> > Can you please take a look at `socket_read()` too ? Should this function always succeed and read the requested bytes when called ?
> 
> It should respect the configured timeout, so by default is blocking and wait forever for at least one byte (or disconnection). It doesn't need to read all the requested bytes, just at least one byte and at most the requested number of bytes.

For what it's worth, I've traced the issue with `webrepl` and `socket_read` to `uos_dupstream.c`. The `socket_read()` and `socket_write()` functions must return `MP_EWOULDBLOCK` or `MP_EAGAIN` on timeout, Not `MP_ETIMEDOUT`, otherwise `uos_dupstream.c` would raise an error and close the stream:
```C
} else if (out_sz == MP_STREAM_ERROR) {
    // errcode is valid
    if (mp_is_nonblocking_error(errcode)) {
        nlr_pop();
    } else {
        mp_raise_OSError(errcode);
    }
```
With `mp_is_nonblocking_error` defined as:
```C
#define mp_is_nonblocking_error(errno) ((errno) == MP_EAGAIN || (errno) == MP_EWOULDBLOCK)
```

This is similar to `modlwip.c` which returns also note the `webrepl` client socket becomes non-blocking after the handshake (see `extmod/webrepl.py:accept_conn`)

dpgeorge commented:
There might need to be some specific tests for this non-blocking read behaviour.  There are already tests for non-blocking accept and connect (see `tests/net_hosted/`).
